### PR TITLE
Moves functionality from gitrepo to gitjob controller

### PR DESCRIPTION
**Pull request by 0xavi0**
Moves the following from `gitrepo` controller to `gitjob` controller:
* AuthorizeAndAssignDefaults
* RBAC resources creation
* Helm secrets check
* NewTargetsConfigMap

The `gitrepo` controller still purges `bundles` and `bundledeployments` on `gitrepo` deletion and handles the status coming from `bundles` and `bundledeployments`

Some functionality that was shared between both controllers has been moved to a common package to avoid code repetition.

Refers to: https://github.com/rancher/fleet/issues/2435

### CODE REVIEW

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (aruiz14) [inline charts/fleet/templates/rbac_gitjob.yaml:56]: Are these really being used?

Comment (0xavi0) [reply charts/fleet/templates/rbac_gitjob.yaml]: you're right... those 2 are not needed... ( :star: the wonders of copy/paste :star2: ). Thanks for catching it!

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (weyfonk) [inline charts/fleet/templates/rbac_gitjob.yaml:58]: Wouldn't this be enough?
```suggestion
      - "create"
```

File: integrationtests/gitjob/controller/controller_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/controller_test.go:255]: nit:
```suggestion
	When("creating a gitRepo that references a nonexistent helm secret", func() {
```

File: integrationtests/gitjob/controller/controller_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/controller_test.go:287]: nit: this would probably also work (same remark about L327):
```suggestion
					return strings.Contains(strLogs, `failed to look up HelmSecretNameForPaths, error: Secret \"secret-does-not-exist\" not found`)
```

File: integrationtests/gitjob/controller/controller_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/controller_test.go:301]: nit:
```suggestion
```

File: integrationtests/gitjob/controller/suite_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/suite_test.go:74]: Nice one :rocket:

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go
Comment (weyfonk) [inline internal/cmd/controller/gitops/reconciler/gitjob_controller.go:203]: Is this comment still necessary, given the name of the method? ;)

**Review by weyfonk — COMMENTED** [review]
This almost LGTM, pending clarification on needed verbs for permissions on roles and role bindings. The rest is mostly nitpicks.
Thanks for this :)

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (0xavi0) [reply charts/fleet/templates/rbac_gitjob.yaml:58]: I tried that, but it was not enough. 
I even tried adding verbs one by one. 
I get the following error, I guess because later it tries to grand RBAC permissions that are different to "create".

```bash
Reconciler error	{"controller": "gitrepo", "controllerGroup": "fleet.cattle.io", "controllerKind": "GitRepo", "GitRepo": {"name":"sample","namespace":"fleet-local"}, "namespace": "fleet-local", "name": "sample", "reconcileID": "2ad237ee-5d4a-49c9-9cb9-b38b1273ceb4", "error": "roles.rbac.authorization.k8s.io \"git-sample\" is forbidden: user \"system:serviceaccount:cattle-fleet-system:gitjob\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:cattle-fleet-system\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held:\n{APIGroups:[\"fleet.cattle.io\"], Resources:[\"bundles\"], Verbs:[\"get\" \"create\" \"update\" \"list\" \"delete\"]}\n{APIGroups:[\"fleet.cattle.io\"], Resources:[\"imagescans\"], Verbs:[\"get\" \"create\" \"update\" \"list\" \"delete\"]}"}
``` 

I even tried with 
* create
* delete
* deletecollection
* get
* list
* patch
* update
* watch 

and I got the same error.

This is still a mistery to me.
I ended up adding `*` because the original RBAC had exactly the same, more than happy to hold this PR to discuss this and understand why.

File: integrationtests/gitjob/controller/controller_test.go
Comment (0xavi0) [reply integrationtests/gitjob/controller/controller_test.go]: changed as suggested

File: integrationtests/gitjob/controller/controller_test.go
Comment (0xavi0) [reply integrationtests/gitjob/controller/controller_test.go]: changed as suggested

File: integrationtests/gitjob/controller/controller_test.go
Comment (0xavi0) [reply integrationtests/gitjob/controller/controller_test.go]: line deleted

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go
Comment (0xavi0) [reply internal/cmd/controller/gitops/reconciler/gitjob_controller.go]: :bowtie: good point. comment deleted

File: integrationtests/gitjob/controller/controller_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/controller_test.go:323]: Shouldn't this be:
```suggestion
			It("logs an error about HelmSecretName not being found", func() {
```
?

File: integrationtests/gitjob/controller/controller_test.go
Comment (weyfonk) [inline integrationtests/gitjob/controller/controller_test.go:326]:
```suggestion
					return strings.Contains(strLogs, `failed to look up HelmSecretName, error: Secret \"secret-does-not-exist\" not found`)
```

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (weyfonk) [reply charts/fleet/templates/rbac_gitjob.yaml:58]: Maybe something like this then?
```
    verbs:
      -  'get'
      - 'create'
      - 'update'
      - 'list'
      - 'delete'
```

Comment (0xavi0) [reply charts/fleet/templates/rbac_gitjob.yaml:58]: I tried that combination as well and I got the same error. :confused:

File: integrationtests/gitjob/controller/controller_test.go
Comment (0xavi0) [reply integrationtests/gitjob/controller/controller_test.go]: Thanks for catching this copy/paste evil thing!

The test was passing because that was the previous value checked and was still in the logs buffer. 
I fixed the name and error message check AND I'm cleaning the `logsBuffer` in the `AfterEach` statement for those tests so the logs are clean and we don't get entries from previous tests.

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (0xavi0) [reply charts/fleet/templates/rbac_gitjob.yaml]: As discussed offline, I had to add `escalate` and `bind` to the `roles`.
It ended up like this:
```yaml
  - apiGroups:
      - ""
    resources:
      - serviceaccounts
    verbs:
      - "create"
  - apiGroups:
      - rbac.authorization.k8s.io
    resources:
      - roles
    verbs:
      - escalate
      - create
      - bind
  - apiGroups:
      - rbac.authorization.k8s.io
    resources:
      - rolebindings
    verbs:
      - create
```

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go
Comment (weyfonk) [inline internal/cmd/controller/gitops/reconciler/gitjob_controller.go:113]: nit: in the same way that we format errors when failing to create a git job (or even better: using the `%w` [placeholder](https://pkg.go.dev/fmt#Errorf) instead of `%v`), we could format errors happening when failing to create RBAC resources or the configmap, to make them easier to understand for users, eg:
```suggestion
			return ctrl.Result{}, fmt.Errorf("failed to create RBAC resources for git job: %w", err)
```

**Review by weyfonk — APPROVED** [review]

File: charts/fleet/templates/rbac_gitjob.yaml
Comment (weyfonk) [reply charts/fleet/templates/rbac_gitjob.yaml]: Nice find :+1: Thanks!

File: integrationtests/gitjob/controller/suite_test.go
Comment (0xavi0) [reply integrationtests/gitjob/controller/suite_test.go]: Oh, yeah, I forgot to change that back to `ctrl`. 
There was already a variable around called `ctrl` and they were clashing.

It should be back to the standard `ctrl` now

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go
Comment (0xavi0) [reply internal/cmd/controller/gitops/reconciler/gitjob_controller.go]: yup, I've changed as suggested, and also... as I didn't want to have `%v`s and `%w`s in the same file... I ended up changing all `%v`s.

**Review by weyfonk — APPROVED** [review]

**Pull request merged**
